pkg/webhook/infrastructure/mutate.go

```diff
@@ -0,0 +1,137 @@
+	var (
+		newInfra, oldInfra *extensionsv1alpha1.Infrastructure
+		ok                 bool
+	)
+
+	if new == nil || old == nil {
+		return nil
+	}
+
+	if new.GetDeletionTimestamp() != nil {
+		return nil
+	}
+
+	oldInfra, ok = old.(*extensionsv1alpha1.Infrastructure)
+	if !ok {
+		return fmt.Errorf("could not mutate: object is not of type Infrastructure")
+	}
+	newInfra, ok = new.(*extensionsv1alpha1.Infrastructure)
+	if !ok {
+		return fmt.Errorf("could not mutate: object is not of type Infrastructure")
+	}
+
+	return m.mutateFunc(ctx, m.logger, newInfra, oldInfra)
+}
+
+// NetworkLayoutMigrationMutate annotates the infrastructure object with additonal information that are necessary during the reconciliation when migrating to a new network layout.
+func NetworkLayoutMigrationMutate(ctx context.Context, logger logr.Logger, newInfra, oldInfra *extensionsv1alpha1.Infrastructure) error {
+	var (
+		newProviderCfg, oldProviderCfg *azure.InfrastructureConfig
+		oldProviderStatus              *azure.InfrastructureStatus
+		err                            error
+	)
+
+	if newInfra.Spec.ProviderConfig == nil {
+		return nil
+	}
+
+	newProviderCfg, err = helper.InfrastructureConfigFromInfrastructure(newInfra)
+	if err != nil {
+		return fmt.Errorf("could not mutate object: %v", err)
+	}
+
+	// if the new one has the label then check if you need to remove it
+	if z, ok := newInfra.Annotations[azuretypes.NetworkLayoutZoneMigrationAnnotation]; ok {
+		findMatchingZone := false
+		for _, zone := range newProviderCfg.Networks.Zones {
+			if helper.InfrastructureZoneToString(zone.Name) == z {
+				findMatchingZone = true
+			}
+		}
+
+		if !findMatchingZone {
+			delete(newInfra.Annotations, azuretypes.NetworkLayoutZoneMigrationAnnotation)
+		}
+		return nil
+	}
+
+	if oldInfra.Spec.ProviderConfig == nil {
+		return nil
+	}
+
+	oldProviderCfg, err = helper.InfrastructureConfigFromInfrastructure(oldInfra)
+	if err != nil {
+		return fmt.Errorf("could not mutate object: %v", err)
+	}
+
+	if oldInfra.Status.ProviderStatus != nil {
+		oldProviderStatus, err = helper.InfrastructureStatusFromInfrastructure(oldInfra)
+		if err != nil {
+			return fmt.Errorf("could not mutate object: %v", err)
+		}
+	}
+
+	// if the new configuration is not zoned or is not using multiple-subnet layout it is not eligible for the mutation.
+	if !newProviderCfg.Zoned || len(newProviderCfg.Networks.Zones) == 0 {
+		return nil
+	}
+
+	// if the old configuration is not zoned and it is already using multiple-subnet layout, no mutation is necessary.
+	if !oldProviderCfg.Zoned || len(oldProviderCfg.Networks.Zones) > 0 {
+		return nil
+	}
+
+	// take care of clusters that have not been reconciliated for a long time (hibernated etc). In this case they may
+	// not have the Layout field populated, so instead rely on the old spec. If the cluster was not a zoned one skip it.
+	if oldProviderStatus != nil &&
+		oldProviderStatus.Networks.Layout != "" &&
```

Comment:
Makes sense, I had the check arranged in ta different manner but I forgot to update the comment. 

For your question, what the code is doing here, is checking "if the `layout` has a value and if it is not the expected one for the mutation" -> terminate early. Otherwise continue with the mutation.